Agent Version-Check is not using configured proxy

**Describe the bug**
We are trying to use the V7-Agent and fail due to a version check problem:
```
$ ./discovery_agent --envFile ./da_env_vars.env
{"level":"trace","msg":"GET [0ms] - ERR - http://0.0.0.0:8989/status - Get \"http://0.0.0.0:8989/status\": dial tcp 0.0.0.0:8989: connect: connection refused","time":"2021-09-29T12:18:38+02:00"}
{"level":"trace","msg":"Approval mode set: manual","time":"2021-09-29T12:18:38+02:00"}
{"level":"trace","msg":"Get cached token is empty.  Try and fetch a new token","time":"2021-09-29T12:18:38+02:00"}
{"level":"debug","msg":"no password, assuming unencrypted key","time":"2021-09-29T12:18:38+02:00"}
{"level":"trace","msg":"Unable to poll jfrog for agent versions. Get \"https://axway.jfrog.io/artifactory/ampc-public-docker-release/agent/v7-discovery-agent/\": read tcp 10.248.0.116:39106-\u003e160.50.180.63:443: read: connection reset by peer","time":"2021-09-29T12:18:38+02:00"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x9c9f09]

goroutine 7 [running]:
github.com/Axway/agent-sdk/pkg/cmd.loadPage(0xb8c201, 0x12, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/!axway/agent-sdk@v1.1.3/pkg/cmd/agentversionjob.go:199 +0x109
github.com/Axway/agent-sdk/pkg/cmd.(*AgentVersionCheckJob).getJFrogVersions(0xc00002c240, 0xb8c201, 0x12, 0x23, 0xc0000eb748)
        /go/pkg/mod/github.com/!axway/agent-sdk@v1.1.3/pkg/cmd/agentversionjob.go:106 +0x39
github.com/Axway/agent-sdk/pkg/cmd.(*AgentVersionCheckJob).Execute(0xc00002c240, 0xc000334068, 0xbbd418)
        /go/pkg/mod/github.com/!axway/agent-sdk@v1.1.3/pkg/cmd/agentversionjob.go:75 +0x2b2
github.com/Axway/agent-sdk/pkg/jobs.(*baseJob).executeJob(0xc000334000)
        /go/pkg/mod/github.com/!axway/agent-sdk@v1.1.3/pkg/jobs/baseJob.go:41 +0x38
github.com/Axway/agent-sdk/pkg/jobs.(*baseJob).start(0xc000334000)
        /go/pkg/mod/github.com/!axway/agent-sdk@v1.1.3/pkg/jobs/baseJob.go:136 +0x5e
created by github.com/Axway/agent-sdk/pkg/jobs.newBaseJob
        /go/pkg/mod/github.com/!axway/agent-sdk@v1.1.3/pkg/jobs/baseJob.go:36 +0x130
```

We have configured a proxy, but it looks like this version check is ignoring the proxy configuration. At least, I don't see anything in this code:
https://github.com/Axway/agent-sdk/blob/998cbcec797b68de00d8b22a2775850b9ea9e673/pkg/cmd/agentversionjob.go#L199-L213

**To Reproduce**
Steps to reproduce the behavior:
- Try to start the discovery agent with Proxy-Only internet access.

**Expected behavior**
The version check routing must use the configured proxy.

--- a/pkg/cmd/agentversionjob.go
+++ b/pkg/cmd/agentversionjob.go
@@ -1,27 +1,29 @@
 package cmd
 
 import (
-	"bytes"
-	"encoding/xml"
+	"encoding/json"
 
 	"github.com/Axway/agent-sdk/pkg/config"
+	"github.com/Axway/agent-sdk/pkg/util"
 
-	"fmt"
-	"io/ioutil"
 	"regexp"
 	"strconv"
 	"strings"
 
 	"net/http"
 
+	coreapi "github.com/Axway/agent-sdk/pkg/api"
 	"github.com/Axway/agent-sdk/pkg/jobs"
 	"github.com/Axway/agent-sdk/pkg/util/errors"
 	log "github.com/Axway/agent-sdk/pkg/util/log"
 )
 
-const avcCronSchedule = "@daily"
+const (
+	avcCronSchedule = "@daily"
+	jfrogURL        = "https://axway.jfrog.io/ui/api/v1/ui/treebrowser"
+)
 
-var agentURL = map[string]string{
+var agentNameToRepoPath = map[string]string{
 	"AWSDiscoveryAgent":                      "aws-apigw-discovery-agent",
 	"AWSTraceabilityAgent":                   "aws-apigw-traceability-agent",
 	"AzureDiscoveryAgent":                    "azure-discovery-agent",
@@ -30,22 +32,86 @@ var agentURL = map[string]string{
 	"EnterpriseEdgeGatewayTraceabilityAgent": "v7-traceability-agent",
 }
 
-type htmlAnchors struct {
-	VersionList []string `xml:"body>pre>a"`
-}
-
 type version struct {
 	major, minor, patch int
 	val                 string
 }
 
+type jfrogRequest struct {
+	Type     string `json:"type"`
+	RepoType string `json:"repoType"`
+	RepoKey  string `json:"repoKey"`
+	Path     string `json:"path"`
+	Text     string `json:"text"`
+}
+
+type jfrogData struct {
+	Items []jfrogItem `json:"data"`
+}
+
+type jfrogItem struct {
+	RepoKey      string `json:"repoKey,omitempty"`
+	Path         string `json:"path,omitempty"`
+	Version      string `json:"text"`
+	RepoType     string `json:"repoType,omitempty"`
+	HasChild     bool   `json:"hasChild,omitempty"`
+	Local        bool   `json:"local,omitempty"`
+	Type         string `json:"type,omitempty"`
+	Compacted    bool   `json:"compacted,omitempty"`
+	Cached       bool   `json:"cached,omitempty"`
+	Trash        bool   `json:"trash,omitempty"`
+	Distribution bool   `json:"distribution,omitempty"`
+}
+
 // AgentVersionCheckJob - polls for agent versions
 type AgentVersionCheckJob struct {
 	jobs.Job
-	allVersions   []string
-	buildVersion  string
-	dataPlaneType string
-	urlName       string
+	apiClient    coreapi.Client
+	requestBytes []byte
+	buildVersion string
+	headers      map[string]string
+}
+
+// NewAgentVersionCheckJob - creates a new agent version check job structure
+func NewAgentVersionCheckJob(cfg config.CentralConfig) (*AgentVersionCheckJob, error) {
+	// get current build version
+	buildVersion, err := getBuildVersion()
+	if err != nil {
+		log.Trace(err)
+		return nil, err
+	}
+
+	if _, found := agentNameToRepoPath[BuildAgentName]; !found {
+		err := errors.ErrStartingVersionChecker.FormatError("empty or generic data plane type name")
+		log.Trace(err)
+		return nil, err
+	}
+
+	// create the request body for each check
+	requestBody := jfrogRequest{
+		Type:     "junction",
+		RepoType: "virtual",
+		RepoKey:  "ampc-public-docker-release",
+		Path:     "agent/" + agentNameToRepoPath[BuildAgentName],
+		Text:     agentNameToRepoPath[BuildAgentName],
+	}
+	requestBytes, err := json.Marshal(requestBody)
+	if err != nil {
+		log.Trace(err)
+		return nil, err
+	}
+
+	return &AgentVersionCheckJob{
+		apiClient:    coreapi.NewClientWithTimeout(cfg.GetTLSConfig(), cfg.GetProxyURL(), cfg.GetClientTimeout()),
+		buildVersion: buildVersion,
+		requestBytes: requestBytes,
+		headers: map[string]string{
+			"X-Requested-With": "XMLHttpRequest",
+			"Host":             "axway.jfrog.io",
+			"Content-Length":   strconv.Itoa(len(requestBytes)),
+			"Content-Type":     "application/json",
+		},
+	}, nil
 }
 
 // Ready -
@@ -60,19 +126,7 @@ func (avj *AgentVersionCheckJob) Status() error {
 
 // Execute - run agent version check job one time
 func (avj *AgentVersionCheckJob) Execute() error {
-	avj.dataPlaneType = BuildAgentName
-	avj.urlName = agentURL[avj.dataPlaneType]
-	if avj.urlName == "AgentSDK" || avj.urlName == "" {
-		err := errors.ErrStartingVersionChecker.FormatError("empty or generic data plane type name")
-		log.Trace(err)
-		return err
-	}
-	err := avj.getBuildVersion()
-	if err != nil {
-		log.Trace(err)
-		return err
-	}
-	err = avj.getJFrogVersions(avj.urlName)
+	err := avj.getJFrogVersions()
 	if err != nil {
 		log.Trace(err)
 		// Could not get update from jfrog.  Warn that we could not determine version and continue processing
@@ -83,21 +137,6 @@ func (avj *AgentVersionCheckJob) Execute() error {
 			log.Warnf("New version available. Please consider upgrading from version %s to version %s", avj.buildVersion, config.AgentLatestVersion)
 		}
 	}
-
-	return nil
-}
-
-func (avj *AgentVersionCheckJob) getBuildVersion() error {
-	avj.buildVersion = BuildVersion
-	//remove -SHA from build version
-	noSHA := strings.Split(avj.buildVersion, "-")
-	avj.buildVersion = noSHA[0]
-
-	//regex check for semantic versioning
-	semVerRegexp := regexp.MustCompile(`\d.\d.\d`)
-	if avj.buildVersion == "" || !semVerRegexp.MatchString(avj.buildVersion) {
-		return errors.ErrStartingVersionChecker.FormatError("build version is missing or of noncompliant semantic versioning")
-	}
 	return nil
 }
 
@@ -105,21 +144,38 @@ func (avj *AgentVersionCheckJob) getBuildVersion() error {
 // **Note** polling the jfrog website is the current solution to obtaining the list of versions
 // In the future, adding a (Generic) resource for grouping versions together under the same scope is a possible solution
 // ie: a new unscoped resource that represents the platform services, so that other products can plug in their releases.
-func (avj *AgentVersionCheckJob) getJFrogVersions(name string) error {
-	b, err := loadPage(name)
+func (avj *AgentVersionCheckJob) getJFrogVersions() error {
+	request := coreapi.Request{
+		Method:  http.MethodPost,
+		URL:     jfrogURL,
+		Headers: avj.headers,
+		Body:    avj.requestBytes,
+	}
+	response, err := avj.apiClient.Send(request)
 	if err != nil {
 		return err
 	}
 
-	hAnchors := htmlAnchors{}
-	err = xml.NewDecoder(bytes.NewBuffer(b)).Decode(&hAnchors)
+	jfrogResponse := jfrogData{}
+	err = json.Unmarshal(response.Body, &jfrogResponse)
 	if err != nil {
 		return err
 	}
 
-	avj.allVersions = hAnchors.VersionList
-	config.AgentLatestVersion = avj.getLatestVersionFromJFrog()
+	config.AgentLatestVersion = avj.getLatestVersionFromJFrog(jfrogResponse.Items)
 	return nil
+}
+
+func getBuildVersion() (string, error) {
+	//remove -SHA from build version
+	versionNoSHA := strings.Split(BuildVersion, "-")[0]
+
+	//regex check for semantic versioning
+	semVerRegexp := regexp.MustCompile(`\d.\d.\d`)
+	if versionNoSHA == "" || !semVerRegexp.MatchString(versionNoSHA) {
+		return "", errors.ErrStartingVersionChecker.FormatError("build version is missing or of noncompliant semantic versioning")
+	}
+	return versionNoSHA, nil
 }
 
 // isVersionStringOlder - return true if version of str1 is older than str2
@@ -146,7 +202,7 @@ func isVersionSmaller(v1 version, v2 version) bool {
 	return false
 }
 
-func (avj *AgentVersionCheckJob) getLatestVersionFromJFrog() string {
+func (avj *AgentVersionCheckJob) getLatestVersionFromJFrog(jfrogItems []jfrogItem) string {
 	tempMaxVersion := version{
 		major: 0,
 		minor: 0,
@@ -155,11 +211,10 @@ func (avj *AgentVersionCheckJob) getLatestVersionFromJFrog() string {
 	}
 	re := regexp.MustCompile(`\d{8}`)
 
-	for _, v := range avj.allVersions {
-		//trimming version of / that comes from jfrog webpage
-		trimmed := strings.TrimSuffix(v, "/")
-		if trimmed != ".." && trimmed != "latest" && trimmed != "" {
-			v := getSemVer(trimmed)
+	for _, item := range jfrogItems {
+		//trimming version from jfrog webpage
+		if item.Version != "latest" && item.Version != "" {
+			v := getSemVer(item.Version)
 			// avoid a version with an 8 digit date as the patch number: 1.0.20210421
 			if !re.MatchString(strconv.Itoa(v.patch)) && isVersionSmaller(tempMaxVersion, v) {
 				copyVersionStruct(&tempMaxVersion, v)
@@ -196,39 +251,21 @@ func copyVersionStruct(v1 *version, v2 version) {
 	v1.val = v2.val
 }
 
-func loadPage(name string) ([]byte, error) {
-	page := fmt.Sprintf("https://axway.jfrog.io/artifactory/ampc-public-docker-release/agent/%v/", name)
-	resp, err := http.Get(page)
-	if err != nil {
-		log.Tracef("Unable to poll jfrog for agent versions. %s", err.Error())
-		return nil, err
+// startVersionCheckJobs - starts both a single run and continuous checks
+func startVersionCheckJobs(cfg config.CentralConfig) {
+	if !util.IsNotTest() {
+		return
 	}
-	defer resp.Body.Close()
-	// reads html as a slice of bytes
-	html, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		log.Trace(err)
-	}
-	return html, nil
-}
-
-// startAgentVersionChecker - single run job to check for a newer agent version on jfrog
-func startAgentVersionChecker() {
 	// register the agent version checker single run job
-	id, err := jobs.RegisterSingleRunJobWithName(&AgentVersionCheckJob{}, "Version Check")
+	checkJob, err := NewAgentVersionCheckJob(cfg)
 	if err != nil {
-		log.Errorf("could not start the agent version checker job: %v", err.Error())
+		log.Errorf("could not create the agent version checker: %v", err.Error())
 		return
 	}
-	log.Tracef("registered agent version checker job: %s", id)
-}
-
-// startAgentVersionCheckerSchedule - cron job that checks for a newer agent version on jfrog on a daily basis
-func startAgentVersionCheckerSchedule() {
-	id, err := jobs.RegisterScheduledJobWithName(&AgentVersionCheckJob{}, avcCronSchedule, "Version Check Schedule")
-	if err != nil {
-		log.Errorf("could not start the agent version checker cronjob: %v", err.Error())
-		return
+	if id, err := jobs.RegisterSingleRunJobWithName(checkJob, "Version Check"); err == nil {
+		log.Tracef("registered agent version checker job: %s", id)
 	}
-	log.Tracef("registered agent version checker cronjob: %s", id)
+	if id, err := jobs.RegisterScheduledJobWithName(checkJob, avcCronSchedule, "Version Check Schedule"); err == nil {
+		log.Tracef("registered agent version checker cronjob: %s", id)
+	}
 }

--- a/pkg/cmd/root.go
+++ b/pkg/cmd/root.go
@@ -275,7 +275,7 @@ func (c *agentRootCommand) initConfig() error {
 		}
 	}
 
-	// Check the agent's latest version once at the agent's start time
+	// Start the initial and recurring version check jobs
 	startVersionCheckJobs(c.centralCfg)
 	// Init the healthcheck API
 	hc.HandleRequests()

--- a/pkg/cmd/root.go
+++ b/pkg/cmd/root.go
@@ -197,10 +197,6 @@ func (c *agentRootCommand) initialize(cmd *cobra.Command, args []string) error {
 
 	c.checkStatusFlag()
 	agentsync.SetSyncMode(c.GetProperties())
-	// Check the agent's latest version once at the agent's start time
-	startAgentVersionChecker()
-	// Start a cronjob that regularly checks the agent's latest version and updates the agent resource
-	startAgentVersionCheckerSchedule()
 	return nil
 }
 
@@ -279,6 +275,8 @@ func (c *agentRootCommand) initConfig() error {
 		}
 	}
 
+	// Check the agent's latest version once at the agent's start time
+	startVersionCheckJobs(c.centralCfg)
 	// Init the healthcheck API
 	hc.HandleRequests()
 	return nil

